### PR TITLE
build(client/electron): build cross-platform with zig

### PR DESCRIPTION
**Pull request by fortuna**
I'm using zig to cross-compile. It's SO MUCH EASIER:
https://dev.to/kristoff/zig-makes-go-cross-compilation-just-work-29ho

/cc @daniellacosse @amircybersec

### CODE REVIEW

File: src/electron/README.md [hunk 1]
```diff
@@ -52,3 +52,12 @@ apt update && apt install -y gcc-mingw-w64-x86-64
 To build the _release_ version of Windows installer, you'll also need:
 
 - [Java 8+ Runtime](https://www.java.com/en/download/). This is required for the cross-platform Windows executable signing tool [Jsign](https://ebourg.github.io/jsign/). If you don't need to sign the executables, feel free to skip this.
+
+## Linux
+
+To build for Linux on a macOS, you need to first install the [musl-cross compiler](https://github.com/GregorR/musl-cross).
```
Comment (jyyi1) [inline src/electron/README.md:58]: The `brew` package seems to reference https://github.com/richfelker/musl-cross-make instead. We might need to carefully decide which one to use.

We can build out own, or download one of the pre-built binary here: https://musl.cc/ . (Though musl declares it to be unofficial, but it's hosted under the `musl` domain name)

File: Makefile [hunk 2]
```diff
@@ -44,11 +44,9 @@ LINUX_BUILDDIR=$(BUILDDIR)/linux
...
 linux: $(LINUX_BUILDDIR)/tun2socks
 
-$(LINUX_BUILDDIR)/tun2socks: $(XGO)
-	mkdir -p "$(LINUX_BUILDDIR)/$(IMPORT_PATH)"
-	$(XGO) -ldflags $(XGO_LDFLAGS) --targets=linux/amd64 -dest "$(LINUX_BUILDDIR)" -pkg $(ELECTRON_PKG) .
-	mv "$(LINUX_BUILDDIR)/$(IMPORT_PATH)-linux-amd64" "$@"
-	rm -r "$(LINUX_BUILDDIR)/$(IMPORT_HOST)"
+$(LINUX_BUILDDIR)/tun2socks:
+	mkdir -p "$(@D)"
+	GOOS=linux GOARCH=amd64 CGO_ENABLED=1 CC=x86_64-linux-musl-gcc go build --ldflags=--extldflags=$(XGO_LDFLAGS) -o "$@" $(ELECTRON_PKG)
```
Comment (jyyi1) [inline Makefile:49]: Shall we statically link the musl-libc library using `...-musl-gcc -static ...` as mentioned in [How to Use musl](https://www.musl-libc.org/how.html)? I don't know whether `musl-gcc` without `-static` requires some musl dynamic library like `libmusl.so` to be installed in the final user's machine.

```suggestion
	GOOS=linux GOARCH=amd64 CGO_ENABLED=1 CC="x86_64-linux-musl-gcc -static" go build --ldflags=--extldflags=$(XGO_LDFLAGS) -o "$@" $(ELECTRON_PKG)
```

File: Makefile [hunk 3]
```diff
@@ -57,7 +55,7 @@ windows: $(WINDOWS_BUILDDIR)/tun2socks.exe
 
 $(WINDOWS_BUILDDIR)/tun2socks.exe:
 	mkdir -p "$(@D)"
-	GOOS=windows GOARCH=amd64 CGO_ENABLED=1 CC="x86_64-w64-mingw32-gcc" go build --ldflags=--extldflags=$(XGO_LDFLAGS) -o "$@" $(ELECTRON_PKG)
+	GOOS=windows GOARCH=amd64 CGO_ENABLED=1 CC=x86_64-w64-mingw32-gcc go build --ldflags=--extldflags=$(XGO_LDFLAGS) -o "$@" $(ELECTRON_PKG)
```
Comment (jyyi1) [inline Makefile:58]: Similar to the static linking mentioned above, maybe we should add `-static` option to Windows `CC` as well? Or at least `-static-libgcc` and `-static-libstdc++`. We should test whether the final binary requires any additional DLLs during runtime.

**Review by jyyi1 — COMMENTED** [review]
Thanks! You mentioned podman is too slow, do we still need podman in this PR?

File: src/electron/README.md (hunk 1 above)
Comment (fortuna) [reply src/electron/README.md]: I changed to zig, which provides its own official package.

**Review by fortuna — COMMENTED** [review]
Another look?

File: Makefile [hunk 4]
```diff
@@ -57,15 +55,12 @@ windows: $(WINDOWS_BUILDDIR)/tun2socks.exe
 
 $(WINDOWS_BUILDDIR)/tun2socks.exe:
 	mkdir -p "$(@D)"
-	GOOS=windows GOARCH=amd64 CGO_ENABLED=1 CC="x86_64-w64-mingw32-gcc" go build --ldflags=--extldflags=$(XGO_LDFLAGS) -o "$@" $(ELECTRON_PKG)
+	GOOS=windows GOARCH=amd64 CGO_ENABLED=1 CC=x86_64-w64-mingw32-gcc go build -trimpath -ldflags=--extldflags=$(LDFLAGS) -o "$@" $(ELECTRON_PKG)
```
Comment (jyyi1) [inline Makefile:58]: `zig` also supports `x86_64-windows` as well. But it only supports Windows 10+. We can keep using MinGW-w64 for backward compatibility now, but I'd prefer to use the same toolchain in the future when we decide to drop the support of old Windows.

File: src/electron/README.md [hunk 5]
```diff
@@ -2,19 +2,6 @@
 
 Unlike the Android and Apple clients, the Windows and Linux clients use the Electron framework, rather than Cordova.
 
-You will need [Docker](https://www.docker.com/) installed to build the Electron clients.
```
Comment (jyyi1) [inline src/electron/README.md:5]: removing containers!!! 👍

File: src/electron/README.md [hunk 6]
```diff
@@ -27,28 +14,21 @@ To run the Electron clients, run:
...
 
-With [MacPorts](https://www.mingw-w64.org/downloads/#macports) (official channel):
+[Install zig](https://ziglang.org/learn/getting-started/#installing-zig) and make sure it's in the PATH.
 
-```sh
-sudo port install x86_64-w64-mingw32-gcc @11.0.1
-```
-
-With Homebrew (unofficial, how to ensure consistent version?):
+You can download the binary tarball, or use a package manager, like Homebrew:
```
Comment (jyyi1) [inline src/electron/README.md:23]: A link to the supported package managers may be useful. It covers almost all platforms including Windows, macOS and Ubuntu.

```suggestion
You can download the binary tarball, or use a [package manager](https://github.com/ziglang/zig/wiki/Install-Zig-from-a-Package-Manager), like Homebrew:
```

File: src/electron/README.md [hunk 7]
```diff
@@ -27,28 +14,21 @@ To run the Electron clients, run:
...
 
 ```sh
-brew install mingw-w64
```
Comment (jyyi1) [inline src/electron/README.md:43]: We might still need to include `mingw-w64` in README since we are still using it?

**Review by jyyi1 — COMMENTED** [review]
The code LGTM, with a few comments! By the way, `zig` seems to be very powerful:

> When it comes to macOS, Zig even has a custom built linker able to cross-compile for both Intel and Apple Silicon M1, something that not even lld (LLVM's linker) can do.

The only concern is `zig` has not published any stable release yet, the latest version is `0.11.0`.

File: src/electron/README.md (hunk 7 above)
Comment (fortuna) [reply src/electron/README.md:43]: I've migrated everything to zig

**Review by fortuna — COMMENTED** [review]
Comments addressed

**Review by jyyi1 — APPROVED** [review]
Thanks a lot! Now we have a more consistent toolchain for build.

**Pull request merged**